Atomic.md

```diff
@@ -0,0 +1,253 @@
+        overwritten by a retry.
+
+-   We crash before reaching InitializeEmptyPlaceholder
+
+    -   We have a partially created placeholder written into .staging that
+        will be overwritten on retry
+
+-   One of the SetBitInFileFlags calls fails in
+     InitializeEmptyPlaceholder
+
+    -   Directory is marked as in virtualization root but not empty.
+
+        -   An incomplete placeholder is in .staging that can be safely
+            overwritten/deleted when the enumeration is retried
+
+-   Chmod() fails or is never reached due to a crash
+
+    -   We can retry and have this succeed as the placeholder hasn't
+        left .staging yet
+
+    -   The rename() to copy the completed placeholder from .staging into
+        the virtualization root fails
+
+        -   We can retry from this error since nothing ever appeared in
+            the virtualization root. The placeholder will be overwritten
+            as part of the retry
+
+With this approach, each placeholder file that will be created will be moved in one at a time to their final location rather than moving all placeholders that may be present in a directory at once. 
+
+In a case where we crash (or the machine crashes, or the user kills the mount process) while expanding a placeholder directory, we will be able to recover by calling UpdatePlaceholderInformation on files that were enumerated by the previous attempt and creating the rest of the placeholders as expected. 
+
+We need to call UpdatePlaceholderInformation in order to prevent data loss in scenarios where users manage to create files in the partially enumerated directory (if the kext is unmounted, the user can write files into the directory that will look like full files to VFSForGit).
+UpdatePlaceholderInformation currently will add any file to ModifiedPaths (which will make 'git status' begin tracking it, if any changes are present) and will one day be smart enough to only update files that are still placeholders.
+
+Designs where we moved an entire directory's worth of placeholders were considered, but were tabled in favor of this approach (making any expansion/hydration retriable).
+
+All of the potential holes identified above are filled by this design,
+so I believe this to be our best bet to create placeholders in an atomic
+matter.
+
+Miscellanea/Things to Validate
+------------------------------
+
+-   Is it possible to hydrate a placeholder in a directory that isn't
+    fully enumerated? It shouldn't be, will revalidate experimentally.
+
+-   Placeholders that are being created in .staging should be given a
+    temporary name (we'll rename() into the actual name). If we give
+    them the exact name of the final file, there is a hole where some
+    issue occurs, they get abandoned in the .staging directory. Once that
+    happens, we'll encounter issues on a retry where we try creating a
+    file where one exists with the same name.
```

Comment:
I'm having trouble understanding this hole.  

I could see wanting to use temp names in case multiple folders are being expanded in parallel, but why couldn't we just overwrite existing files in .staging (if needed)?

Reply:
That's a better justification for this hole, I will switch to that explanation.